### PR TITLE
APP-183 [C] Endpoint profiles and grouping

### PR DESCRIPTION
**Pull request by Dubland**
/cc @RostakaGmfun @vadimol

### CODE REVIEW

File: profilingdemo/build.sh [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    help
+fi
+
+APP_NAME="demo_client"
+PROJECT_HOME=$(pwd)
+BUILD_DIR="build"
+KAA_TOOLCHAIN_PATH_SDK=""
+
+build_app() {
+    cd $PROJECT_HOME
```
Comment (vadimol) [inline profilingdemo/build.sh]: use quotes : "$PROJECT_HOME", the PROJECT_HOME can contain spaces. 
Or better, use [shellcheck](https://www.shellcheck.net/) for all spells

File: profilingdemo/source/c/src/kaa_demo.c [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    if (strcmp(parameter, "+") == 0) {
+        return true;
+    }
+    return false;
+}
+
+int main(int argc, char *argv[])
+{   
+    if (argc != 5) {
+        printf("Few arguments\n");
```
Comment (vadimol) [inline profilingdemo/source/c/src/kaa_demo.c]: use demo_printf() or printf(), but not both

File: profilingdemo/source/c/src/kaa_demo.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        return true;
+    }
+    return false;
+}
+
+int main(int argc, char *argv[])
+{   
+    if (argc != 5) {
+        demo_printf("Few arguments\n");
+        exit(1);
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c:71]: Use `EXIT_FAILURE`.

File: profilingdemo/source/c/src/kaa_demo.c [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    mkdir(argv[4], 0777);
+    chdir(argv[4]);
+
+    /**
+     * Initialize Kaa client.
+     */
+    kaa_error_t error_code = kaa_client_create(&kaa_client, NULL);
+    if (error_code) {
+        demo_printf("Failed create Kaa client, error code %d\n", error_code);
+        return error_code;
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: `EXIT_FAILURE`

File: profilingdemo/source/c/src/kaa_demo.c [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <kaa_error.h>
+#include <platform/kaa_client.h>
+#include <kaa_configuration_manager.h>
+#include <kaa_profile.h>
+#include <platform/ext_key_utils.h>
+
+#define demo_printf(msg, ...) printf((msg), ##__VA_ARGS__)
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: The `demo_printf` macro is defined in `common/c/targets/...`.
If you prefer to use it (and if demo is planned to be run on non-posix platforms), include the `common/c` directory into the demo.

File: profilingdemo/source/c/src/kaa_demo.c [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            NULL,
+            &kaa_demo_configuration_receiver,
+    };
+
+    error_code = kaa_configuration_manager_set_root_receiver(
+            kaa_client_get_context(kaa_client)->configuration_manager,
+            &receiver);
+    if (error_code) {
+        demo_printf("Failed to add configuration receiver\r\n");
+        return error_code;
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: `EXIT_FAILURE`

File: profilingdemo/source/c/src/kaa_demo.c [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+            kaa_configuration_manager_get_configuration(
+                kaa_client_get_context(kaa_client)->configuration_manager));
+
+    /**
+     * Start Kaa client main loop.
+     */
+    error_code = kaa_client_start(kaa_client, NULL, NULL, 0);
+    if (error_code) {
+        demo_printf("Failed to start Kaa main loop, error code %d\n", error_code);
+        return error_code;
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: `EXIT_FAILURE`

File: profilingdemo/source/c/src/kaa_demo.c [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    
+    /**
+     * Destroy Kaa client.
+     */
+    kaa_client_destroy(kaa_client);
+
+    return error_code;
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: `EXIT_SUCCESS`

File: profilingdemo/source/c/src/kaa_demo.c [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    };
+
+    error_code = kaa_configuration_manager_set_root_receiver(
+            kaa_client_get_context(kaa_client)->configuration_manager,
+            &receiver);
+    if (error_code) {
+        demo_printf("Failed to add configuration receiver\r\n");
+        return error_code;
+    }
+    demo_printf("Configuration\n");
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: Please rewrite this message to more self-explanatory one

File: profilingdemo/source/c/src/kaa_demo.c [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+
+    demo_printf("Profiling body (have audio-, vibro-, video-support):\n");
+    demo_printf("%s - %s - %s\n", profile->audio_support ? "true" : "false",
+                                  profile->vibro_support ? "true" : "false",
+                                  profile->video_support ? "true" : "false");
+
+    error_code = kaa_profile_manager_update_profile(kaa_client_get_context(kaa_client)->profile_manager, profile);
+    if (error_code) {
+        demo_printf("Failed to update profile\r\n");
+        return error_code;
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: `EXIT_FAILURE`

File: profilingdemo/source/c/src/kaa_demo.c (hunk 4 above)
Comment (vadimol) [inline profilingdemo/source/c/src/kaa_demo.c]: resource leak, kaa_cloient must be cleaned.

File: profilingdemo/source/c/src/kaa_demo.c [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    kaa_error_t error_code = kaa_client_create(&kaa_client, NULL);
+    if (error_code) {
+        demo_printf("Failed create Kaa client, error code %d\n", error_code);
+        return error_code;
+    }
+    
+    kaa_profile_t *profile = kaa_profile_pager_client_profile_create();
+    if (!profile) {
+        demo_printf("Failed to create profile\r\n");
+        return EXIT_FAILURE;
```
Comment (vadimol) [inline profilingdemo/source/c/src/kaa_demo.c:90]: resource leak

File: profilingdemo/source/c/src/kaa_demo.c [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    
+    const uint8_t *endpoint_key_hash = NULL;
+    size_t endpoint_key_hash_length = 0;
+
+    ext_get_sha1_base64_public(&endpoint_key_hash, &endpoint_key_hash_length);
+
+    demo_printf("- - -\n");
+    demo_printf("Endpoint Key Hash %.*s\r\n", (int)endpoint_key_hash_length,
```
Comment (RostakaGmfun) [inline profilingdemo/source/c/src/kaa_demo.c]: Don't print `\r`

File: profilingdemo/source/c/src/kaa_demo.c [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    if (strcmp(parameter, "+") == 0) {
+        return true;
+    }
+    return false;
+}
+
+int main(int argc, char *argv[])
+{   
+    if (argc != 5) {
+        demo_printf("Few arguments\n");
```
Comment (vadimol) [inline profilingdemo/source/c/src/kaa_demo.c]: what if argc > 5?

File: profilingdemo/source/c/src/kaa_demo.c [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        exit(1);
+    }
+    
+    mkdir(argv[4], 0777);
+    chdir(argv[4]);
```
Comment (vadimol) [inline profilingdemo/source/c/src/kaa_demo.c:75]: why do you need this?

Comment (Dubland) [inline profilingdemo/source/c/src/kaa_demo.c:75]: For creating kaaClient with different keyHash

**Comment by RostakaGmfun** [comment]
Please remove `profilingdemo/source/c/src/.kaa_demo.c.swp`.

**Comment by Slash32** [comment]
LGTM

**Pull request merged**